pep-0544.txt

```diff
@@ -0,0 +1,949 @@
+      ...
+  class GoodJob:
+      ...
+      def quit(self) -> int:
+          return 0
+  finish(GoodJob()) # OK
+
+In addition, we propose to add another special type construct
+``All`` that represents intersection types. Although for normal types
+it is not very useful, there are many situations where a variable should
+implement more than one protocol. Annotation by ``All[Proto1, Proto2, ...]``
+means that a given variable or parameter must implement all protocols
+``Proto1``, ``Proto2``, etc. either implicitly or explicitly. Example::
+
+  from typing import Sequence, Hashable, All
+
+  def cached_func(args: All[Sequence[float], Hashable]) -> float:
+      ...
+  cached_func((1, 2, 3)) # OK, tuple is hashable and sequence
+
+The interaction between union and intersection types is specified by PEP 483,
+and basically reflects the corresponding interactions for sets.
+
+
+``Type[]`` with protocols
+-------------------------
+
+Variables and parameters annotated with ``Type[Proto]`` accept only concrete
+(non-protocol) subtypes of ``Proto``. For example::
+
+  class Proto(Protocol):
+      @abstractmethod
+      def meth(self) -> int:
+          ...
+  class Concrete:
+      def meth(self) -> int:
+          return 42
+
+  def fun(cls: Type[Proto]) -> int:
+      return cls().meth() # OK
+  fun(Proto)              # Error
+  fun(Concrete)           # OK
+
+The same rule applies to variables::
+
+  var: Type[Proto]
+  var = Proto    # Error
+  var = Concrete # OK
+  var().meth()   # OK
+
+Assigning a protocol class to a variable is allowed if it is not explicitly
+typed, and such assignment creates a type alias. For non-protocol classes,
+the behavior of ``Type[]`` is not changed.
+
+
+``NewType()`` and type aliases
+------------------------------
+
+Protocols are essentially anonymous. To emphasize this point, static type
+checkers might refuse protocol classes inside ``NewType()`` to avoid an
+illusion that a distinct type is provided::
+
+  form typing import NewType , Protocol, Iterator
+
+  class Id(Protocol):
+      code: int
+      secrets: Iterator[bytes]
+
+  UserId = NewType('UserId', Id)  # Error, can't provide distinct type
+
+On the contrary, type aliases are fully supported, including generic type
+aliases::
+
+  from typing import TypeVar, Reversible, Iterable, Sized
+
+  T = TypeVar('T')
+  CompatReversible = Union[Reversible[T], All[Iterable[T], Sized]]
+
+
+.. _discussion:
+
+Narrowing types by ``isinstance()``
+-----------------------------------
+
+Protocol types will implement special instance and class checks, similar to
+how ``Iterable`` and other ABCs in ``collections.abc`` and ``typing`` already
+do it, but this is limited to non-generic and unsubscripted generic protocols
+(``Iterable`` is statically equivalent to ``Iterable[Any]`).
+The metaclass of ``typing.Protocol`` will automatically add
+a ``__subclasshook__()`` that provides the same semantics
+for class and instance checks as for ``collections.abc`` classes::
```

Comment:
I don't love the automatic `isinstance` check support, even though that's how the mypy `Protocol` class used to work. I don't think that they can be made to work reliably for arbitrary ABCs. Consider this example:

```py
class P(Protocol):
    @abstractmethod
    def common_method_name(self, x: int) -> int: ... 

class X:
    <a bunch of methods>
    def common_method_name(self) -> None: ...

def do_stuff(o: Union[P, X]) -> int:
    if isinstance(o, P):
        return o.common_method_name(1)  # oops what if it's an X instance instead?
    else:
        return (do something with an X)
```

Maybe a type checker can warn about this example, but more generally the `isinstance` check could be unreliable, except for things where there is a common signature convention such as `Iterable`.

Another potentially problematic case:

```py
class P(Protocol):
    x: int

class C:
    def initialize(self) -> None:
        self.x = 0

c1 = C()
c2 = C()
c2.initialize()
isinstance(c1, P)  # False
isinstance(c2, P)  # True

def f(x: Union[P, int]) -> None:
    if isinstance(x, P):
        # type of x is P here
        ...
    else:
        # type of x is int here?
        print(x + 1)

f(C())   # oops
```

Again, maybe we can live with this, but this could be pretty surprising. I'd argue that requiring an explicit class decorator would be better, since we can then attach warnings about problems like this in the documentation. The programmer would be able to evaluate whether the benefits outweigh the potential for confusion for each protocol and explicitly opt in -- but the default behavior would be safer.